ta/crypto_perf/ta_crypto_perf.c

```diff
@@ -1,291 +1,494 @@
+			CHECK(res, "pre-flush out memref param", return res;);
+		}
+	}
+
+	in = params[0].memref.buffer;
+	insz = params[0].memref.size;
+	out = params[1].memref.buffer;
+	outsz = params[1].memref.size;
+	n = params[2].value.a;
+	unit = params[2].value.b;
+	if (!unit)
+		unit = insz;
+
+	if (algo == TEE_ALG_AES_GCM)
+		do_update = TEE_AEUpdate;
+	else
+		do_update = TEE_CipherUpdate;
+
+	while (n--) {
+		uint32_t i = 0;
+		for (i = 0; i < insz / unit; i++) {
+			res = do_update(crypto_op, in, unit, out, &outsz);
+			CHECK(res, "TEE_CipherUpdate/TEE_AEUpdate", return res;);
+			in  = (void *)((uintptr_t)in + unit);
+			out = (void *)((uintptr_t)out + unit);
+		}
+		if (insz % unit) {
+			res = do_update(crypto_op, in, insz % unit, out, &outsz);
+			CHECK(res, "TEE_CipherUpdate/TEE_AEUpdate", return res;);
+		}
+	}
+
+	if (secure_out) {
+		/* intentionally flush output data from cache for SDP buffers */
+		res = flush_memref_buffer(&params[1]);
+		CHECK(res, "post-flush out memref param", return res;);
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result cmd_cipher_prepare_key(uint32_t param_types, TEE_Param params[4])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_ObjectHandle hkey = TEE_HANDLE_NULL;
+	TEE_ObjectHandle hkey2 = TEE_HANDLE_NULL;
+	TEE_ObjectType objectType;
+	TEE_Attribute attr = { };
+	uint32_t mode = 0;
+	uint32_t op_keysize = 0;
+	uint32_t keysize = 0;
+	const uint8_t *ivp = NULL;
+	size_t ivlen = 0;
+	static uint8_t cipher_key[] = { 0x00, 0x01, 0x02, 0x03,
+				     0x04, 0x05, 0x06, 0x07,
+				     0x08, 0x09, 0x0A, 0x0B,
+				     0x0C, 0x0D, 0x0E, 0x0F,
+				     0x10, 0x11, 0x12, 0x13,
+				     0x14, 0x15, 0x16, 0x17,
+				     0x18, 0x19, 0x1A, 0x1B,
+				     0x1C, 0x1D, 0x1E, 0x1F };
+	static uint8_t cipher_key2[] = { 0x20, 0x21, 0x22, 0x23,
+				      0x24, 0x25, 0x26, 0x27,
+				      0x28, 0x29, 0x2A, 0x2B,
+				      0x2C, 0x2D, 0x2E, 0x2F,
+				      0x30, 0x31, 0x32, 0x33,
+				      0x34, 0x35, 0x36, 0x37,
+				      0x38, 0x39, 0x3A, 0x3B,
+				      0x3C, 0x3D, 0x3E, 0x3F };
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mode = params[0].value.a ? TEE_MODE_DECRYPT : TEE_MODE_ENCRYPT;
+	keysize = params[0].value.b;
+	op_keysize = keysize;
+
+	switch (params[1].value.a) {
+	case TA_AES_ECB:
+		algo = TEE_ALG_AES_ECB_NOPAD;
+		objectType = TEE_TYPE_AES;
+		use_iv = 0;
+		break;
+	case TA_AES_CBC:
+		algo = TEE_ALG_AES_CBC_NOPAD;
+		objectType = TEE_TYPE_AES;
+		use_iv = 1;
+		break;
+	case TA_AES_CTR:
+		algo = TEE_ALG_AES_CTR;
+		objectType = TEE_TYPE_AES;
+		use_iv = 1;
+		break;
+	case TA_AES_XTS:
+		algo = TEE_ALG_AES_XTS;
+		objectType = TEE_TYPE_AES;
+		use_iv = 1;
+		op_keysize *= 2;
+		break;
+	case TA_AES_GCM:
+		algo = TEE_ALG_AES_GCM;
+		objectType = TEE_TYPE_AES;
+		use_iv = 1;
+		break;
+	case TA_SM4_ECB:

```

Comment:
SM4 support looks new, please mention that in the commit message.

Reply:
OK，I will delete SM4-concerned changes and create another pull request later.